### PR TITLE
Added localization framework to mssql-cli

### PR DESCRIPTION
**Pull request by geleems**
Added localization framework to mssql-cli.

### CODE REVIEW

File: mssqlcli/locale/mssql-cli.pot [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Translations template for PROJECT.
```
Comment (pensivebrian) [inline mssqlcli/locale/mssql-cli.pot]: Why are we checking in the .pot? Isn't this a template to generate the .po files?

Comment (geleems) [reply mssqlcli/locale/mssql-cli.pot]: I followed how azdata did. You can see azdata repository contains *.pot files generated by pybabel command.

File: mssqlcli/mssqlclioptionsparser.py [hunk 2]
```diff
@@ -15,139 +19,139 @@
 def create_parser():
     args_parser = argparse.ArgumentParser(
         prog=u'mssql-cli',
-        description=u'Microsoft SQL Server CLI. ' +
-        u'Version {}'.format(mssqlcli.__version__))
+        description=_(u'Microsoft SQL Server CLI Version {}')\
```
[the remark this answers is not in the document]
Comment (geleems) [reply mssqlcli/mssqlclioptionsparser.py]: fixed.

File: mssqlcli/completion_refresher.py [hunk 3]
```diff
@@ -56,7 +59,7 @@ def _bg_refresh(self, mssqlcliclient, callbacks, history=None,
         if not owner_uri:
             # If we were unable to connect, do not break the experience for the user.
             # Return nothing, smart completion can maintain the keywords and functions completions.
-            logger.error(u'Completion refresher connection failure.'.join(error_messages))
+            logger.error(_(u'Completion refresher connection failure').join(error_messages))
```
[the remark this answers is not in the document]
Comment (geleems) [reply mssqlcli/completion_refresher.py]: fixed.

File: mssqlcli/completion_refresher.py [hunk 4]
```diff
@@ -40,7 +43,7 @@ def refresh(self, mssqcliclient, callbacks, history=None,
             self._completer_thread.setDaemon(True)
             self._completer_thread.start()
             return [(None, None, None,
-                     'Auto-completion refresh started in the background.')]
+                    _('Auto-completion refresh started in the background.'))]
```
Comment (pensivebrian) [inline mssqlcli/completion_refresher.py]: Is this a log message?

File: mssqlcli/completion_refresher.py [hunk 5]
```diff
@@ -31,7 +34,7 @@ def refresh(self, mssqcliclient, callbacks, history=None,
         """
         if self.is_refreshing():
             self._restart_refresh.set()
-            return [(None, None, None, 'Auto-completion refresh restarted.')]
+            return [(None, None, None, _('Auto-completion refresh restarted.'))]
```
Comment (pensivebrian) [inline mssqlcli/completion_refresher.py]: Is this a log message?

File: mssqlcli/completion_refresher.py (hunk 4 above)
Comment (geleems) [reply mssqlcli/completion_refresher.py]: fixed.

File: mssqlcli/completion_refresher.py (hunk 5 above)
Comment (geleems) [reply mssqlcli/completion_refresher.py]: fixed.

File: mssqlcli/locale/mssql-cli.pot (hunk 1 above)
Comment (geleems) [reply mssqlcli/locale/mssql-cli.pot]: Removed pot from commit

File: mssqlcli/mssqlclioptionsparser.py [hunk 6]
```diff
@@ -15,139 +19,139 @@
...
-        help=u'If not supplied, defaults to value in environment variable MSSQL_CLI_PASSWORD.')
+        help=_(u'If not supplied, defaults to value in environment variable MSSQL_CLI_PASSWORD.'))
 
     args_parser.add_argument(
         u'-d', u'--database',
         dest=u'database',
         default=os.environ.get(MSSQL_CLI_DATABASE, None),
         metavar=u'',
-        help=u'database name to connect to.')
+        help=_(u'Database name to connect to'))
```
Comment (pensivebrian) [inline mssqlcli/mssqlclioptionsparser.py]: What' happened to the period?

Comment (geleems) [reply mssqlcli/mssqlclioptionsparser.py]: Sentence should end with period. Noun should not.

Comment (pensivebrian) [reply mssqlcli/mssqlclioptionsparser.py]: Then let's fix the sentence. The convention in this file is to use a period for every help string.

File: mssqlcli/i18n.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from importlib import import_module
+import gettext
+import os
+
+CLI_NAME = 'mssql-cli'
+MODULE = 'mssqlcli'
+
+def translate():
```
Comment (pensivebrian) [inline mssqlcli/i18n.py]: Should we use gettext.install?
[https://docs.python.org/3/library/gettext.html#gettext.install](https://docs.python.org/3/library/gettext.html#gettext.install)

File: mssqlcli/mssqlclioptionsparser.py (hunk 6 above)
Comment (geleems) [reply mssqlcli/mssqlclioptionsparser.py]: > Then let's fix the sentence.

I could not exactly understand what it means. My terminology 'sentence' means a set of words that is complete in itself, typically starts with subject + verb.
* "If not supplied, defaults to value in environment variable MSSQL_CLI_PASSWORD." --> This is a sentence.
* "database name to connect to." --> This is not a sentence, and should not have period at the end as my knowledge of English grammar.

I do not see any consistent convention both in this file and across the project. For example, "Username to connect to the database" does not have period while "database name to connect to." does have. Other string such as "Connection request encountered an exception" in connectionservice.py does not have period.

I rolled back all period change I made in this PR since it seems going out of scope. Let's have the string convention in separate PR later.

File: mssqlcli/i18n.py (hunk 7 above)
Comment (geleems) [reply mssqlcli/i18n.py]: The project is built and runs successfully without additional installation for gettext since it is built-in module already included in default python.

Comment (pensivebrian) [reply mssqlcli/i18n.py]: Invoking `gettext.install()` will install the _() function to translate string. This way, you don't have to add the `_ = translate()` code in the modules needing translation.

File: mssqlcli/i18n.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    Gets the localized translation of message, based on the current global
+    domain, language, and locale directory of the module.
+    :param mod: The module namespace containing the locale for translation.
+    :return: The localized translation of message, based on the current global
+             domain, language, and locale directory alias.
+    """
+    modulePath = os.path.dirname(import_module(MODULE).__file__)
+    localedir = os.path.join(modulePath, 'locale')
+    domain = CLI_NAME
+    return gettext.translation(domain, localedir, fallback=True).gettext
```
Comment (pensivebrian) [inline mssqlcli/i18n.py]: I think having a per module locale directory is still more than we need. Can we have a single 'locale' directory at the root? And then add the following to top of main.py?

```
import gettext
translation = gettext.translation('mssql-cli', localedir='locale', fallback=True)
translation.install()
```

**Comment by pensivebrian** [comment]
The are some strings that need to be translated in main.py

**Comment by pensivebrian** [comment]
Same with mssql_cli.py. Are you planning on making a complete loc string pass in another PR?

File: mssqlcli/i18n.py (hunk 8 above)
Comment (geleems) [reply mssqlcli/i18n.py]: fixed.

File: mssqlcli/main.py [hunk 9]
```diff
@@ -1,21 +1,25 @@
...
-from mssqlcli.config import config_location
-from mssqlcli.__init__ import __version__
-from mssqlcli.mssqlclioptionsparser import create_parser
-import mssqlcli.telemetry as telemetry_session
-
 click.disable_unicode_literals_warning = True
 
+translation = gettext.translation('mssql-cli', localedir='locale', fallback=True)
+translation.install()
+_ = translation.gettext
```
Comment (pensivebrian) [inline mssqlcli/main.py]: You shouldn't need this line - doesn't install take care of that?

**Comment by pensivebrian** [comment]
Is it possible to add a unit test to make sure all of this works?

File: mssqlcli/main.py (hunk 9 above)
Comment (geleems) [reply mssqlcli/main.py]: I don't understand. Did you mean `_` method is automatically defined by `translation.install()` ?

Comment (pensivebrian) [reply mssqlcli/main.py]: Correct. See https://docs.python.org/3/library/gettext.html#gettext.install

File: mssqlcli/i18n.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
+import gettext
+import os
+from importlib import import_module
+import utility
+import polib
```
Comment (pensivebrian) [inline mssqlcli/i18n.py]: We cannot import babel defined modules in this file since babel is a dev dependency and not a runtime dependency. Let's move all babel related code into build.py

File: mssqlcli/jsonrpc/contracts/connectionservice.py [hunk 11]
```diff
@@ -4,6 +4,9 @@
```
Comment (pensivebrian) [inline mssqlcli/jsonrpc/contracts/connectionservice.py]: Let's undo this changes to this file since we're only localizing the 'goodbye!' message.

File: mssqlcli/mssqlclioptionsparser.py [hunk 12]
```diff
@@ -11,143 +11,143 @@
 MSSQL_CLI_ROW_LIMIT = u'MSSQL_CLI_ROW_LIMIT'
 MSSQL_CLI_RC = u'MSSQL_CLI_RC'
 
+from mssqlcli.i18n import translation
```
Comment (pensivebrian) [inline mssqlcli/mssqlclioptionsparser.py]: Let's undo this changes to this file since we're only localizing the 'goodbye!' message.

File: tests/test_localization.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            'extraction_target': i18n.MODULE_ROOT_DIR,
+            'lang_name': 'ko_KR',
+            'domain': i18n.DOMAIN,
+            'trans_mappings': {
+                u'Goodbye!': u'안녕히가세요!',
+            }
+        }
+        i18n.generate_mo(mo_config)
+
+        i18n.LANGUAGES = ['ko']
```
Comment (pensivebrian) [inline tests/test_localization.py]: Let's check in the .mo file with localized 'goodbye!' string. This way it's picked by the unit tests. Once we have the lcl->mo translation process in place, we'll remove this .mo file since they will be generated as part of the build process.

File: tests/test_localization.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        base_dir = mo_config['base_dir']
+        localedir = os.path.join(base_dir, 'locale')
+        lang_name = mo_config['lang_name']
+        mo_dir = os.path.join(localedir, lang_name, 'LC_MESSAGES')
+        self.create_dir([base_dir, 'locale', lang_name, 'LC_MESSAGES'])
+
+        try:
+            import babel
+        except ImportError:
+            utility.exec_command('pip install --upgrade babel', utility.ROOT_DIR)
```
[the remark this answers is not in the document]
Comment (pensivebrian) [reply tests/test_localization.py]: My concerns are around runtime a dependency. Dev is fine.

File: build.py [hunk 15]
```diff
@@ -169,6 +171,52 @@ def validate_actions(user_actions, valid_targets):
             sys.exit(1)
 
 
+def generate_mo(mo_config):
+    extraction_target = getDictValue(mo_config, 'extraction_target')
```
Comment (pensivebrian) [inline build.py]: This method accepts a dictionary, which is very generic. Can it accept actual parameters so that callers can more easily discover and use the method correctly?

File: mssqlcli/localized_strings.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+def translation(domain=DOMAIN, localedir=LOCALE_DIR, languages=None):
+    languages = languages if (not languages is None) else LANGUAGES
+    return gettext.translation(domain, localedir, languages, fallback=True)
+
+translation().install()
+
+
+## Localized Strings
+def goodbye():
+    return _('Goodbye!')
```
Comment (pensivebrian) [inline mssqlcli/localized_strings.py]: what about the u prefix for strings?

File: mssqlcli/mssql_cli.py [hunk 17]
```diff
@@ -87,6 +85,7 @@ def security_words_found_in(query):
         return False
 
 
+# pylint: disable=bad-super-call
```
Comment (pensivebrian) [inline mssqlcli/mssql_cli.py]: Let's revert all pylint suppressions. We probably should fix these issues, and should do so holistically in another PR.

File: mssqlcli/mssqlclioptionsparser.py [hunk 18]
```diff
@@ -15,8 +15,7 @@
 def create_parser():
     args_parser = argparse.ArgumentParser(
         prog=u'mssql-cli',
-        description=u'Microsoft SQL Server CLI. ' +
-        u'Version {}'.format(mssqlcli.__version__))
+        description=u'Microsoft SQL Server CLI. Version {}'.format(mssqlcli.__version__))
```
Comment (pensivebrian) [inline mssqlcli/mssqlclioptionsparser.py]: Do we need to change this file? If not, let's revert. We have lots of flake8 and formatting issues but let's address those in other PRs.

File: mssqlcli/localized_strings.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+import gettext
+import os
+from importlib import import_module
+
+PATH = '{0}.py'.format(os.path.splitext(__file__)[0])
+DOMAIN = 'mssql-cli'
+LOCALE_DIR = os.path.join(os.path.dirname(__file__), 'locale')
+LANGUAGES = None
+
+def translation(domain=DOMAIN, localedir=LOCALE_DIR, languages=None):
```
Comment (pensivebrian) [inline mssqlcli/localized_strings.py]: Can we validate the language items such that we throw if passed and invalid language?

File: tests/test_localization.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        translated = self.unicode(localized.goodbye())
+        assert original != translated
+
+    def test_ko_KR(self):
+        domain, ext = os.path.splitext(os.path.basename(__file__))
+        mo_config = {
+            'extraction_target': '{0}.py'.format(os.path.splitext(__file__)[0]),
+            'lang_name': 'ko_KR',
+            'domain': domain,
+            'trans_mappings': {
```
Comment (pensivebrian) [inline tests/test_localization.py]: Can we checkin the .mo file with the translated 'goodbye!' message? This way, we're testing the same logic the product code is using?

File: build.py (hunk 15 above)
Comment (geleems) [reply build.py]: This method is created for only for testing. I would stay with dictionary type since it is easier to add or remove properties. Can we just put comment defining argument spec?

File: mssqlcli/localized_strings.py (hunk 16 above)
Comment (geleems) [reply mssqlcli/localized_strings.py]: fixed.

File: mssqlcli/mssqlclioptionsparser.py (hunk 18 above)
Comment (geleems) [reply mssqlcli/mssqlclioptionsparser.py]: reverted

File: mssqlcli/localized_strings.py (hunk 19 above)
Comment (geleems) [reply mssqlcli/localized_strings.py]: will tackle in separate PR as we discussed offline.

File: tests/test_localization.py (hunk 20 above)
Comment (geleems) [reply tests/test_localization.py]: discussed offline

File: mssqlcli/mssql_cli.py (hunk 17 above)
Comment (geleems) [reply mssqlcli/mssql_cli.py]: reverted

File: build.py (hunk 15 above)
Comment (geleems) [reply build.py]: fixed as discussed.

File: mssqlcli/mssqlclioptionsparser.py [hunk 21]
```diff
@@ -120,8 +120,7 @@ def create_parser():
         dest=u'multi_subnet_failover',
         action=u'store_true',
         default=False,
-        help=u'If application is connecting to AlwaysOn AG on different subnets, setting this provides faster '
-             u'detection and connection to currently active server.')
+        help=u'If application is connecting to AlwaysOn AG on different subnets, setting this provides faster detection and connection to currently active server.')
```
Comment (pensivebrian) [inline mssqlcli/mssqlclioptionsparser.py]: Let's revert this change as it's not related to loc.

File: mssqlcli/localized_strings.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+    languages = languages if (not languages is None) else LANGUAGES
+    return gettext.translation(domain, localedir, languages, fallback=True)
+
+translation().install()
+
+
+## Localized Strings
+def goodbye():
+    return _(u'Goodbye!')
```
Comment (pensivebrian) [inline mssqlcli/localized_strings.py]: Add a newline?

File: build.py [hunk 23]
```diff
@@ -169,6 +171,44 @@ def validate_actions(user_actions, valid_targets):
             sys.exit(1)
 
 
+def generate_mo(extraction_target_path, lang_name, trans_mappings, domain, localedir=None):
```
Comment (pensivebrian) [inline build.py]: We should separate out the po and mo generation, but we can do this in another PR.
